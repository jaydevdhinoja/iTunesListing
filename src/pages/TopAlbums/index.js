import React, { useState } from "react";

// Components
import Grid from "../../components/Grid";
import Card from "../../components/Card";
import LoadMoreBtn from "../../components/elements/LoadMoreBtn";
import Spinner from "../../components/elements/Spinner/Spinner";
import TopSongsSection from "../../components/TopSongsSection";

//Global Context
import { useGlobalListingContext } from "../../ContextProvider"

const TopAlbums = () => {
  const {albums, songs, loading, error, favourites, addToFavourite} =  useGlobalListingContext()
  const filterItems = 20;
  const [visible, setVisible] = useState(filterItems);
  const [search, setSearch] = useState("");

  const filteredAlbums = albums.filter(album => {
    return (
      album["im:name"].label.toLowerCase().indexOf(search.toLowerCase()) !== -1 ||
      album["im:artist"].label.toLowerCase().indexOf(search.toLowerCase()) !== -1
    );
  });

  const loadMoreAlbums = () => {
    setVisible(prev => prev + filterItems);
  };

  if (error) return <div>Something went wrong...</div>;

  return (
    <>
      <Grid header="Top Albums" callback={s => setSearch(s)}>
        {filteredAlbums.slice(0, visible).map(album => (
          <Card
            key={album.id.attributes["im:id"]}
            item={album}
            index={albums.indexOf(album)}
            callback={addToFavourite}
            itemType={'albums'}
            favourite={favourites['albums'].filter(item => item === album.id.attributes["im:id"]).length > 0}
            showDetails={true}
          />
        ))}
      </Grid>

      {loading && <Spinner />}

      {visible < filteredAlbums.length && !loading && (
        <LoadMoreBtn text={`Load more albums`} callback={loadMoreAlbums} />
      )}

      {songs.length !== 0 && !loading && <TopSongsSection songs={songs.slice(0, 8)} />}
    </>
  );
};

export default TopAlbums;
