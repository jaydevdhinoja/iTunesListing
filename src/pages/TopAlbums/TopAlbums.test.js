import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { renderWithContext, InitialState } from '../../tests/common';
import TopAlbums from '.';

describe('Page Listing', () => {
	test('renders component', () => {
		const { container } = renderWithContext(<TopAlbums />, { value: InitialState });

		expect(container).toMatchSnapshot();
	});

	test('renders component with no albums', () => {
		const { queryAllByText } = renderWithContext(<TopAlbums />, {
			value: { ...InitialState, albums: [] }
		});

		expect(queryAllByText('Black Pumas')).toHaveLength(0);
	});

	test('renders component with no songs', () => {
		const { queryAllByText } = renderWithContext(<TopAlbums />, {
			value: { ...InitialState, songs: [] }
		});

		expect(queryAllByText('Tim McGraw & Tyler Hubbard')).toHaveLength(0);
	});
	
	test('renders top albums', () => {
		const { getByText } = renderWithContext(<TopAlbums />, { value: InitialState });
		expect(getByText('Black Pumas')).toBeInTheDocument();
	});

	test('renders top songs', () => {
		const { getByText } = renderWithContext(<TopAlbums />, { value: InitialState });
		expect(getByText('Tim McGraw & Tyler Hubbard')).toBeInTheDocument();
	});
});
