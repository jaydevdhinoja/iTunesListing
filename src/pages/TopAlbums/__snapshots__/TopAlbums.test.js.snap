// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Listing renders component 1`] = `
<div>
  <section
    class="albums-grid container"
  >
    <div
      class="albums-grid__headline section-headline"
    >
      <h2>
        Top Albums
      </h2>
      <div
        class="albums-grid__headline-search"
      >
        <input
          class="search-input"
          placeholder="Search"
          type="text"
          value=""
        />
        <img
          alt="Search"
          src="search-solid.svg"
        />
      </div>
    </div>
    <ul
      class="albums-grid__list"
    >
      <li
        class="item-card"
      >
        <a
          href="/item/1524099463"
        >
          <div
            class="item-card__mobile-link"
          />
        </a>
        <div
          class="item-card__thumbnail"
          style="background-image: url(undefined);"
        >
          <img
            alt="Heart Icon"
            class="favourite-icon"
            src="heart-solid.svg"
          />
        </div>
        <h3
          class="item-card__title"
        >
          Black Pumas (Deluxe)
        </h3>
        <p
          class="item-card__artist"
        >
          Black Pumas
        </p>
        <div
          class="item-card__overlay"
        />
        <p
          class="item-card__overlay-number"
        >
          1
        </p>
        <a
          href="/albums/1524099463"
        >
          <span
            class="item-card__overlay-rect"
          >
            <p>
              See albums
            </p>
          </span>
        </a>
        <span
          class="item-card__overlay-shadow"
        />
      </li>
    </ul>
  </section>
  <section
    class="top-songs"
  >
    <div
      class="container"
    >
      <div
        class="top-songs__header"
      >
        <div
          class="top-songs__header-headline section-headline"
        >
          <h2>
            Top Songs
          </h2>
        </div>
      </div>
      <div
        class="top-songs__list"
      >
        <div
          class="row"
        >
          <div
            class="col-md-3"
          >
            <div
              class="top-songs__item"
            >
              <img
                alt="Song Undivided"
                class="top-songs__item-thumbnail"
              />
              <div
                class="top-songs__item-content"
              >
                <p
                  class="top-songs__item-number"
                >
                  #
                  1
                </p>
                <h3
                  class="top-songs__item-title"
                >
                  Undivided
                </h3>
                <p
                  class="top-songs__item-artist"
                >
                  Tim McGraw & Tyler Hubbard
                </p>
                <p
                  class="top-songs__item-artist"
                >
                  2021
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
